mlrun/package/common.py

```diff
@@ -0,0 +1,164 @@
+        # typing classes are considered a type (like `TypeVar`) so we check their type as well. The only case "types"
+        # will be a module is for generic aliases like `list[int]`.
+        return (type_hint.__module__ == "typing") or (
+            type(type_hint).__module__ in ["typing", "types"]
+        )
+
+    @staticmethod
+    def reduce_type_hint(
+        type_hint: typing.Union[typing.Type, typing.Set[typing.Type]],
+    ) -> typing.Set[typing.Type]:
+        """
+        Reduce a type hint (or a list of type hints) using the `_reduce_type_hint` function.
+
+        :param type_hint: The type hint to reduce.
+
+        :return: The reduced type hints set or an empty set if the type hint could not be reduced.
+        """
+        # Wrap in a list if provided a single type hint:
+        type_hints = {type_hint} if not isinstance(type_hint, set) else type_hint
+
+        # Iterate over the type hints and reduce each one:
+        return set(
+            list(
+                itertools.chain(
+                    *[
+                        TypingUtils._reduce_type_hint(type_hint=type_hint)
+                        for type_hint in type_hints
+                    ]
+                )
+            )
+        )
+
+    @staticmethod
+    def _reduce_type_hint(type_hint: typing.Type) -> typing.List[typing.Type]:
+        """
+        Reduce a type hint. If the type hint is a `typing` module, it will be reduced to its original hinted types. For
+        example: `typing.Union[int, float, typing.List[int]]` will return `[int, float, List[int]]` and
+        `typing.List[int]` will return `[list]`. Regular type hints - Python object types cannot be reduced as they are
+        already a core type.
+
+        If a type hint cannot be reduced, an empty list will be returned.
+
+        :param type_hint: The type hint to reduce.
+
+        :return: The reduced type hint as list of hinted types or an empty list if the type hint could not be reduced.
+        """
+        # TODO: Remove when we'll no longer support Python 3.7:
+        if sys.version_info[1] < 8:
+            return []
+
+        # If it's not a typing type (meaning it's an actual object type) then we can't reduce it further:
+        if not TypingUtils.is_typing_type(type_hint=type_hint):
+            return []
+
+        # If it's a type var, take its constraints (e.g. A = TypeVar("A", int, str) meaning an object of type A should
+        # be an integer ot a string). If it doesn't have constraints, return an empty list:
+        if isinstance(type_hint, typing.TypeVar):
+            if len(type_hint.__constraints__) == 0:
+                return []
+            return list(type_hint.__constraints__)
+
+        # If it's a forward reference, that means the user could not import the class to type hint it (so we can't
+        # either):
+        if isinstance(type_hint, typing.ForwardRef):
+            return []
```

Comment:
I'll add something :)

Reply:
`ForwardRef`s will now try to be imported, if they can't we'll throw a warning.
```
[warning] Could not reduce the type hint 'ForwardRef('SomeClass')' as it is a forward reference to a class without it's full module path. To enable importing forward references, please provide the full module path to them. For example: use `ForwardRef('pandas.DataFrame')` instead of `ForwardRef('DataFrame')`.
```